src/chainstate/burn/db/sortdb.rs

```diff
@@ -992,6 +1017,92 @@ impl <'a> SortitionHandleConn <'a> {
         Ok(key_status)
     }
 
+    /// Return a vec of sortition winner's burn header hash and stacks header hash, ordered by
+    ///   increasing block height in the range (block_height_begin, block_height_end]
+    fn get_sortition_winners_in_fork(&self, block_height_begin: u32, block_height_end: u32) -> Result<Vec<(Txid, u64)>,  BurnchainError> {
+        let mut result = vec![];
+        for height in (block_height_begin+1)..(block_height_end+1) {
+            debug!("Looking for winners at height = {}", height);
+            let snapshot = SortitionDB::get_ancestor_snapshot(self, height as u64, &self.context.chain_tip)?
+                .ok_or_else(|| BurnchainError::MissingParentBlock)?;
+            if snapshot.sortition {
+                result.push((snapshot.winning_block_txid, snapshot.block_height));
+            }
+        }
+        Ok(result)
+    }
+
+    pub fn get_reward_cycle_info(&self, prepare_end: &BurnchainHeaderHash, pox_consts: &PoxConstants) -> Result<Option<(ConsensusHash, BlockHeaderHash)>, CoordinatorError> {
+        let prepare_end_sortid = self.get_sortition_id_for_bhh(prepare_end)?
+            .ok_or_else(|| BurnchainError::MissingParentBlock)?;
+        let my_height = SortitionDB::get_block_height(self.deref(), &prepare_end_sortid)?
+            .expect("CORRUPTION: SortitionID known, but no block height in SQL store");
+
+        // there can be either 1 or 0 PoX anchors.
+        assert!(pox_consts.anchor_threshold > (pox_consts.prepare_length / 2));
+
+        // if this block is the _end_ of a prepare phase,
+        if ! (my_height % pox_consts.reward_cycle_length == 0) {
+            return Err(CoordinatorError::NotPrepareEndBlock)
+        }
+
+        let prepare_end = my_height;
+        // if this block isn't greater than prepare_length, then this shouldn't be the end of a prepare block
+        if prepare_end < pox_consts.prepare_length {
+            return Err(CoordinatorError::NotPrepareEndBlock)
+        }
+
+        let prepare_begin = prepare_end - pox_consts.prepare_length;
```

Comment:
Because of the way `prepare_end` and `prepare_begin` are calculated, it looks like this method may be run over _any_ sequence of `pox_consts.prepare_length`-length?  I bring this up because I don't think this method should be called unless `my_height` corresponds to the start of a new reward cycle -- i.e. `my_height % pox_constants.reward_cycle_length == 0` (or at least, `my_height` falls within a registration window, and the caller wants to see what the _current_ anchor block support looks like.).

Reply:
This method will error out with `NotPrepareEndBlock` if this method is called on anything but the end of a prepare block. Otherwise, the calculation being performed here will calculate the PoX anchor of the wrong window.